Conditional Coverage

bh2smith opened the pull request:
Unit testing full Deposit and Withdraw batches using a contract wrapper that enables multi-request submission. Closes #61 

**Test Plan**
```
truffle test
```

bh2smith committed:
test rollover of full deposit and withdraw batches.

fleupold commented on contracts/libraries/wrapppers/MultiRequest.sol:
```diff
@@ -0,0 +1,20 @@
+import "../../SnappBase.sol";
+
+
+contract MultiRequest is SnappBase {
```
nit: This is not really a library, is it? Should we just create a test folder inside `contracts` and place it there? Also, I'd slightly prefer the name `MultiFlux`, but no strong feeling. 

If you wanted to be super cool, you could even make this thing target contract agnostic by exposing a single method that takes the target contract, the encoded calldata (method id and and inputs as byte array) and calls `address.call(calldata)` `multiplier` times.

This might be useful in other projects too.

bh2smith replied:
Yes, I would love to make these suggestions happen, but am not currently aware how. Show me?

fleupold replied:
https://github.com/gnosis/mock-contract#mocking-methods--arguments shows you how you can use `encodeAbi` in js to get the encoded byte array of a method call.

Inside your multi contract you would do something like: 

```
function callMultiple(address contract, bytes calldata) {
  for (i=0, i<multiplier, i++) {
    contract.call(calldata)
  }
}
```

@rmeissner might also be able to help you with specific questions while I'm out.

bh2smith replied:
Awesome. I have started a small repo [here](https://github.com/bh2smith/solidity-multi-transaction) that is now working on functions that don't have any parameters. Will try and get back to this again soon and move it in to gnosis once it's in better shape with more features.

fleupold commented on test/multi_request.js:
```diff
@@ -0,0 +1,55 @@
+    await MultiRequest.link(SnappBaseCore, lib2.address)
+  })
+
+  describe("Test full request batches", () => {
```
Should this rather be part of the existing SnappBase tests? I think it logically fits in there and it feels kind of weird to just have a file for this part of the tests.

bh2smith replied:
Works for me! Will change.

fleupold reviewed (commented):
I'm fine with the MultiRequest contract (although I think you could make this a neat little npm package that other projects could use with a little extra effort).

However, I'd like to hear your thoughts on whether or not the test cases should live in the same file as the existing ones.

bh2smith committed:
Change name of contract to MultiFlux and move unit tests into SnappBase test file

bh2smith commented:
I have changed the name of the file to `MultiFlux` and moved the testing into the appropriate file as suggested. Notice how this contract is inside the `wrappers` subdirectory of `libraries` (since these are used to wrap around contracts). 

Weird thing here. If I take the `MultiFlux` file and move it from `contract/libraries/wrappers/` to `contracts/tests` I get this strange error 

```
Error: Could not find artifacts for MultiFlux.sol from any sources
```

I would be very happy to hear more about making this a proper npm package @fleupold

fleupold commented:
> Weird thing here. If I take the MultiFlux file and move it from contract/libraries/wrappers/ to contracts/tests I get this strange error

Did you maybe forget to `truffle compile` after moving? I can move it into `contracts/test` without a problem.

josojo commented on contracts/libraries/SnappBaseCore.sol:
```diff
@@ -63,6 +63,14 @@ library SnappBaseCore {
     /**
      * Public View Methods
      */
+    function depositBatchSize() public pure returns (uint8) {
```
DEPOSIT_BATCH_SIZE is already declared as public. I guess you could just use the official read function instead of coding your own read function

bh2smith replied:
Yes! Great catch! Somehow, I was not able to make this work yesterday, but it seems fine now!

```js
const batchSize = (await core.DEPOSIT_BATCH_SIZE.call()).toNumber()
```

josojo reviewed (approved):
Looks all good to me!

I guess developing a generalized library would not be worth the effort, as coding it  in a generic way might be a lot of work.

bh2smith commented:
Hey @josojo - Although it might not be worth the effort I have already started trying to build such a generalized tool [here](https://github.com/bh2smith/solidity-multi-transaction)! Check it out! So far it only works on functions that take no parameters.

bh2smith committed:
move MultiFlux into `contracts/test` (as suggested).

bh2smith committed:
No need for helper get functions!

bh2smith commented:
All suggested changes from @fleupold were implemented, so his approval is now assumed.
